torch_xla/distributed/xla_dist.py

```diff
@@ -86,14 +99,34 @@ def __init__(self,
                ' will interfere with the values set for distributed'
                ' training'.format(dist_var)))
 
+  def _check_client_mesh_health(self):
+
+    def _pop_mark_step_queue(queue, timeout=30):
+      queue.get(timeout=timeout)
+
+    try:
+      parallel_work(
+        len(self._heartbeats), _pop_mark_step_queue, self._heartbeats.values())
+    except queue.Empty:
+      self._error_queue.put(RuntimeError(
+        'Client mesh is unhealthy due to non-matching heartbeats'))
+
   def _stream_logs(self, process, client_worker):
-    client_ip = client_worker._internal_ip
-    ordinal = self._cluster._client_workers.index(client_worker)
+    client_ip = client_worker.internal_ip
+    ordinal = self._cluster.client_workers.index(client_worker)
 
     def _stream_output(stream, log_fn):
       for std in iter(stream.readline, b''):
+        std_line = std.decode('utf-8').rstrip('\n')
+        if 'torch_xla.core.xla_model::mark_step' in std_line:
+          self._heartbeats[client_ip].put(1)
+          if client_ip == self._client_master.internal_ip:
```

Comment:
Every time you get a step marker (which happens multiple times per second), you kick a new thread and check health?
The the parallel worker can create another hundreds of threads.

The logic should be different.
If there is an hang, step markers will stop flowing.
So you should have a global "last step marker time", and expensive trigger checks only if `NOW - LAST_STEP_TIME` is greater than a given time delta (say 30s).

Reply:
All threads should've returned ~immediately in healthy case, but yeah other approach sounds good. Done.